src/structure/spatial.jl

```diff
@@ -1,66 +1,298 @@
+    "C"  in atomnames(res)      ? nothing : throw(ArgumentError("Atom with atom name \"C\" not found in residue"))
+    return dihedralangle(res_prev["C"], res["N"], res["CA"], res["C"])
+end
+
+"""
+Calculate the psi angle in radians for an `AbstractResidue`. Arguments are the
+residue (atoms "N", "CA" and "C" required) and the next residue (atom "N"
+required).
+"""
+function psiangle(res::AbstractResidue, res_next::AbstractResidue)
+    "N"  in atomnames(res)      ? nothing : throw(ArgumentError("Atom with atom name \"N\" not found in residue"))
+    "CA" in atomnames(res)      ? nothing : throw(ArgumentError("Atom with atom name \"CA\" not found in residue"))
+    "C"  in atomnames(res)      ? nothing : throw(ArgumentError("Atom with atom name \"C\" not found in residue"))
+    "N"  in atomnames(res_next) ? nothing : throw(ArgumentError("Atom with atom name \"N\" not found in next residue"))
+    return dihedralangle(res["N"], res["CA"], res["C"], res_next["N"])
+end
+
+
+"""
+Calculate the `Vector`s of phi and psi angles of a `StructuralElementOrList`.
+The vectors have `nothing` for residues where an angle cannot be calculated,
+e.g. due to missing atoms or lack of an adjacent residue.
+Additional arguments are residue selector functions - only residues that return
+`true` from the functions are retained.
+"""
+function ramachandranangles(el::StructuralElementOrList,
+                    residue_selectors::Function...)
+    res_list = collectresidues(el, residue_selectors...)
+    if length(res_list) < 2
+        throw(ArgumentError("Multiple residues required to calculate Ramachandran angles"))
+    end
+    phi_angles = Union{Float64, Void}[nothing] # First res has no previous res
+    psi_angles = Union{Float64, Void}[]
+    # Phi angles
+    for i in 2:length(res_list)
+        res = res_list[i]
+        res_prev = res_list[i-1]
+        if sequentialresidues(res_prev, res)
+            try
+                phi_angle = phiangle(res, res_prev)
+                push!(phi_angles, phi_angle)
+            catch ex
+                isa(ex, ArgumentError) || rethrow()
+                push!(phi_angles, nothing)
+            end
+        else
+            push!(phi_angles, nothing)
+        end
+    end
+    # Psi angles
+    for i in 1:length(res_list)-1
+        res = res_list[i]
+        res_next = res_list[i+1]
+        if sequentialresidues(res, res_next)
+            try
+                psi_angle = psiangle(res, res_next)
+                push!(psi_angles, psi_angle)
+            catch ex
+                isa(ex, ArgumentError) || rethrow()
+                push!(psi_angles, nothing)
+            end
+        else
+            push!(psi_angles, nothing)
+        end
+    end
+    push!(psi_angles, nothing) # Last res has no next res
+    return phi_angles, psi_angles
+end
+
+
+"""
+Calculate the contact map for a `StructuralElementOrList`, or between two
+`StructuralElementOrList`s. This is an `Array` with `true` where the
```

Comment:
Actually, it will not be an `Array` object if you use `falses`. It is `BitArray` (or `BitMatrix`).